contracts/apps/accounting/FinanceApp.sol

```diff
@@ -0,0 +1,432 @@
+    @param token Address of the token being transfered
+    @param amount Amount being transfered
+    @param reference custom string to describe the transaction
+    @param _type 0 for deposit 1 for withdrawl
+    @return uint of the new transaction id
+    */
+    function _newTransaction(address externalAddress, address token, uint256 amount, string reference, TransactionType _type) internal returns (uint) {
+        AccountingPeriod memory ap = accountingPeriods[getCurrentAccountingPeriodId()];
+        uint tid = transactions.push(Transaction({
+            token: token,
+            amount: amount,
+            externalAddress: externalAddress,
+            reference: reference,
+            timestamp: now,
+            accountingPeriodId: getCurrentAccountingPeriodId(),
+            _type: _type
+        })) - 1;
+        // All transactions must have at least one state.
+        // To optimize, incoming transactions could go directly to "Suceeded" or "Failed".
+
+        _updateTransaction(tid, TransactionState.New, "new");
+        return tid;
+    }
+
+    /**
+    @dev Approve a pending transaction
+    @param transactionId The ID of the transaction to approve
+    @param reason string of the reason to approve the transaction
+    */
+    function approveTransaction(uint transactionId, string reason) auth external {
+        _approveTransaction(transactionId, reason);
+    }
+
+    function _approveTransaction(uint transactionId, string reason) internal {
+        Transaction memory t = transactions[transactionId];
+        var (state, r) = getTransactionState(transactionId);
+        require(state == TransactionState.PendingApproval);
+        require(t._type == TransactionType.Withdrawal);
+        _setTransactionApproved(transactionId, reason);
+        _executeTransaction(transactionId);
+    }
+
+    /**
+    @dev Deny a pending transaction
+    @param transactionId The ID of the transaction to deny
+    @param reason string of the reason to deny the transaction
+    */
+    function denyTransaction(uint transactionId, string reason) auth external {
+        _denyTransaction(transactionId, reason);
+    }
+
+    function _denyTransaction(uint transactionId, string reason) internal {
+        Transaction memory t = transactions[transactionId];
+        var (state, r) = getTransactionState(transactionId);
+        require(state == TransactionState.PendingApproval);
+        require(t._type == TransactionType.Withdrawal);
+        _setTransactionDenied(transactionId, reason);
+    }
+
+    /**
+    @dev Function to actually transfer the tokens to external address
+    */
+    function _executeTransaction(uint transactionId) internal  {
+        Transaction memory t = transactions[transactionId];
+        var (state, r) = getTransactionState(transactionId);
+        require(state == TransactionState.Approved);
+        bool succeeded = ERC20(t.token).transfer(t.externalAddress, t.amount);
```

Comment:
This needs to be a transferFrom. The FinanceApp doesn't have a balance in the tokens, just whatever it was allowed.